libs/metall/src/dice/metall/detail/file.hpp

```diff
@@ -281,54 +279,165 @@ inline bool copy_file_dense(const std::filesystem::path &source_path,
+ * @param src_size size of file behind src as obtained by ::fstat
+ * @return if copying was successful
+ *
+ * Relevant man pages:
+ *      https://www.man7.org/linux/man-pages/man2/lseek.2.html
+ *      https://www.man7.org/linux/man-pages/man2/copy_file_range.2.html
+ *      https://www.man7.org/linux/man-pages/man3/ftruncate.3p.html
+ */
+inline bool copy_file_sparse_linux(int src, int dst, off_t src_size) {
+  off_t old_off = 0;
+  off_t off = 0;
 
-    bp::ipstream is;
-    bp::child c{"/usr/bin/cp", "--sparse=always", source_path.c_str(), destination_path.c_str(),
-                bp::std_out > bp::null,
-                bp::std_err > is};
+  while ((off = ::lseek(src, off, SEEK_DATA)) >= 0) {
+    if (!create_hole_linux(dst, off - old_off)) {
+      METALL_ERROR("Unable to punch hole");
+      return false;
+    }
 
-    c.wait();
-    if (c.exit_code() == 0) {
-      fsync(destination_path);
-      return true;
+    off_t const hole_start = ::lseek(src, off, SEEK_HOLE);
+    if (hole_start < 0) {
+      METALL_ERRNO_ERROR("fseek(SEEK_HOLE)");
+      return false;
     }
 
-    std::string error;
-    std::copy(std::istreambuf_iterator<char>{is}, std::istreambuf_iterator<char>{}, std::back_inserter(error));
-    METALL_ERROR("Error during sparse copy: {}", error);
-  } catch (...) {
-    METALL_ERROR("Unable to spawn child /usr/bin/cp");
+    if (::copy_file_range(src, &off, dst, nullptr, hole_start - off, 0) < 0) {
+      METALL_ERRNO_ERROR("copy_file_range");
+      return false;
+    }
+
+    old_off = off;
+  }
+
+  if (errno != ENXIO) {
+    // error condition: offset is _not_ within a hole at the end of the file.
+    // previous lseek from while-loop condition must have failed
+    METALL_ERRNO_ERROR("fseek(SEEK_DATA)");
+    return false;
+  }
+
+  if (old_off < src_size) {
+    // the final extent is a hole we must call ftruncate
+    // here in order to record the proper length in the destination.
+    // See also: https://github.com/coreutils/coreutils/blob/a257b63ce7ebcc4577adb5406b39fc0edd61dcac/src/copy.c#L643-L658
+    if (::ftruncate(dst, src_size) < 0) {
+      METALL_ERRNO_ERROR("ftruncate");
+      return false;
+    }
+  }
+
+  if (!create_hole_linux(dst, src_size - old_off)) {
+    METALL_ERROR("Unable to punch hole");
+    return false;
+  }
+
+  return true;
+}
+
+/**
+ * Attempts to perform a sparse copy from source_path to destination_path,
+ * falling back to regular copy if the sparse copy fails.
+ */
+inline bool copy_file_sparse_linux(const std::filesystem::path &source_path,
+                                   const std::filesystem::path &destination_path) {
+  int src;
+  int dst;
+  off_t src_size = prepare_file_copy_linux(source_path, destination_path, &src, &dst);
+  if (src_size < 0) {
+    METALL_ERROR("Unable to prepare for file copy");
+    return false;
+  }
+
+  if (copy_file_sparse_linux(src, dst, src_size)) {
+    os_fsync(dst);
+    os_close(src);
+    os_close(dst);
+    return true;
   }
```

Comment:
is it safe to continue working with src and dst if copy_file_sparse_linux failed? It would be cleaner to close them and run prepare_file_copy_linux again.

Reply:
Thats just a logic error, they should be closed regardless of success because the next step doesn't require them

Reply:
fixed